### DISCUSSION TITLE
Using UnsafeAccessor to access `List<T>._items`

### DISCUSSION THREAD

**Discussion by sfiruch** [dotnet/runtime#110964] [Q&A] [answered]
I'm trying to access the internal `_list` field of `List<T>`. It doesn't appear to work. Here's my experiment on [Sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyUBuauxsgOgCUYAzADYwwGAJYQ8bGgya984gLYwuAYQiKADmOFQAyjCgA3MWBgBnadSMBDKPUFjzGegF56eGAHd6AGScYADxMtAB8ABQAlPQA3vQARDYAFhAAnimp8fQAvlZUAPT5sEIi4pL0XtAA1ubWdvRieBiGeDaCAJLNiuYA8oIAJm704SEA2gC6kY7OXADiMBgAKqmaMFFzCwBiYjAD4fEA+mJd5vFo9ABCjf2NAOabgja35lwAcpIACjjAjmD0BJdrncHk8Xu08M4bHgzJEAIQbDAANTaODW0wwcOkTAAnOFGs0oK0Oic+v0uL4YHhbhgkpE8oUAKoQmz8GAAQTAZnM5mg9H6EAseAA5C4bJpVvUMBAKtUAPx1ez4lptTowbqvbxDeYYfzOVXdcLounsMi4pWElUnDVecmU6m0vLUUZM8ws9mciw8qDhF1ujlcr0AaWuXG2u3651eNmUQ0OxzVp0i42oMAQBLwnCQ9GK9EWE3o2t1GH15kCiwiRbLoQcATpQA=):
```cs
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Runtime.CompilerServices;

var list = new List<string>() { "ahoyhoy" };

//reflection works
var internalItemsOld = (string[])list.GetType().GetField("_items", BindingFlags.NonPublic | BindingFlags.Instance)!.GetValue(list)!;
Console.WriteLine(internalItemsOld.Length);

//UnsafeAccessor doesn't appear to work?
var internalItemsNew = GetListItems(list);
Console.WriteLine(internalItemsNew.Length);


[UnsafeAccessor(UnsafeAccessorKind.Field, Name = "_items")]
extern static ref T[] GetListItems<T>(List<T> list);
```

The reflection-based way works fine but the `UnsafeAccessor` throws this exception:
```
System.MissingFieldException: Field not found: 'System.Collections.Generic.List`1._items'.
```

UnsafeAccessor-with-generics had a bumpy start, but since https://github.com/dotnet/runtime/issues/89439 (.NET 9) it should be working, no? Am I doing something wrong?

**Comment by KeterSCP** [comment, marked as answer]
The answer is the same as in https://github.com/dotnet/runtime/issues/110054#issuecomment-2492908454

In your case, `T` should be type parameter and not method parameter, as in `List<T>` `T` - is a type parameter.
Following code works as expected:

```csharp
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

var list = new List<string>() { "ahoyhoy" };

var internalItemsNew = Accessors<string>.GetListItems(list);
Console.WriteLine(internalItemsNew.Length);

class Accessors<T>
{
    [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "_items")]
    public extern static ref T[] GetListItems(List<T> list);
}
```

**Comment by Clockwork-Muse** [comment]
.... Do you really need the underlying array?  There's also [CollectionsMarshal.AsSpan](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan?view=net-9.0), which would give you a `Span` back.

**Reply by sfiruch** [reply]
Yes, I want to use the same code paths for `List<T>` and `T[]`, and can't use `Span<T>`. I described the use case in https://github.com/dotnet/runtime/issues/105179.

**Reply by hamarb123** [reply]
Just call `AsSpan` once on each thread. It's pretty cheap. And you can use spans in lambdas, you just can't pass them into the lambdas like that as they're not boxable and that's how it gives the state implicitly.

**Reply by Clockwork-Muse** [reply]
... Encapsulate the code in a method/func that takes a `Span`.  You can pass an array directly, so the only "extra" thing you have to do is call `AsSpan` on `List<T>`.  This is valid code on .NET 9+:
```csharp
using System;
					
public class Program
{
	public static void Main()
	{
		int[] n = { 1, 2, 3};
		var r = SomeFuncUser(n, z => z.Length > 1);
		Console.WriteLine(r);
	}
	
	public static bool SomeFuncUser(Span<int> data, Func<Span<int>, bool> something) {
		return something(data);
	}
}
```
